File: Data Structures & Algorithms/binary-tree-diameter/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int d = 0;

    int binaryTreeHeight(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }

        int leftHeight = binaryTreeHeight(root->left);
        int rightHeight = binaryTreeHeight(root->right);
        d = max(d, leftHeight + rightHeight);

        return max(leftHeight, rightHeight) + 1;
    }
    
public:
    int diameterOfBinaryTree(TreeNode* root) {
        d = 0;
        int leftHeight = binaryTreeHeight(root);
        return d;
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sameTree(self, first: Optional[TreeNode], second: Optional[TreeNode]):
        if not first or not second:
            return first == second
        if first.val != second.val:
            return False
        return self.sameTree(first.left, second.left) and self.sameTree(first.right, second.right)
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return root == subRoot
        return self.sameTree(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)